// @enable3d/common/dist/types.js
"use strict";
/**
 * @author       Yannick Deubel (https://github.com/yandeu)
 * @copyright    Copyright (c) 2020 Yannick Deubel; Project Url: https://github.com/enable3d/enable3d
 * @license      {@link https://github.com/enable3d/enable3d/blob/master/LICENSE|GNU GPLv3}
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Group = exports.Points = exports.Line = exports.Material = exports.Mesh = exports.Object3D = exports.AnimationAction = exports.AnimationClip = exports.AnimationMixer = exports.ExtendedMesh = exports.ExtendedObject3D = void 0;
const extendedObject3D_1 = require("./extendedObject3D");
Object.defineProperty(exports, "ExtendedObject3D", { enumerable: true, get: function () { return extendedObject3D_1.ExtendedObject3D; } });
const extendedMesh_1 = require("./extendedMesh");
Object.defineProperty(exports, "ExtendedMesh", { enumerable: true, get: function () { return extendedMesh_1.ExtendedMesh; } });
// TODO fix this export
// export { ThirdPersonControls } from '../misc/thirdPersonControls'
var index_1 = require("@enable3d/three-wrapper/dist/index");
Object.defineProperty(exports, "AnimationMixer", { enumerable: true, get: function () { return index_1.AnimationMixer; } });
Object.defineProperty(exports, "AnimationClip", { enumerable: true, get: function () { return index_1.AnimationClip; } });
Object.defineProperty(exports, "AnimationAction", { enumerable: true, get: function () { return index_1.AnimationAction; } });
Object.defineProperty(exports, "Object3D", { enumerable: true, get: function () { return index_1.Object3D; } });
Object.defineProperty(exports, "Mesh", { enumerable: true, get: function () { return index_1.Mesh; } });
Object.defineProperty(exports, "Material", { enumerable: true, get: function () { return index_1.Material; } });
Object.defineProperty(exports, "Line", { enumerable: true, get: function () { return index_1.Line; } });
Object.defineProperty(exports, "Points", { enumerable: true, get: function () { return index_1.Points; } });
Object.defineProperty(exports, "Group", { enumerable: true, get: function () { return index_1.Group; } });
