// @enable3d/three-graphics/dist/plugins/index.js
"use strict";
/**
 * These Plugins are not included in the Core three-graphics package
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Cameras = exports.HaveSomeFun = exports.WebXR = exports.Transform = exports.TextureCube = exports.Misc = exports.Mixers = exports.WarpSpeed = exports.HeightMap = exports.CSG = exports.Factories = exports.Lights = exports.Loaders = void 0;
const loaders_1 = require("./loaders");
exports.Loaders = loaders_1.default;
const lights_1 = require("./lights");
exports.Lights = lights_1.default;
const factories_1 = require("@enable3d/common/dist/factories");
exports.Factories = factories_1.default;
const csg_1 = require("./csg");
exports.CSG = csg_1.default;
const heightmap_1 = require("./heightmap");
exports.HeightMap = heightmap_1.default;
const warpSpeed_1 = require("./warpSpeed");
exports.WarpSpeed = warpSpeed_1.default;
const mixers_1 = require("./mixers");
exports.Mixers = mixers_1.default;
const misc_1 = require("./misc");
exports.Misc = misc_1.default;
Object.defineProperty(exports, "TextureCube", { enumerable: true, get: function () { return misc_1.TextureCube; } });
const transform_1 = require("./transform");
exports.Transform = transform_1.default;
const webxr_1 = require("./webxr");
exports.WebXR = webxr_1.default;
const haveSomeFun_1 = require("./haveSomeFun");
exports.HaveSomeFun = haveSomeFun_1.default;
const cameras_1 = require("./cameras");
exports.Cameras = cameras_1.default;
