// @enable3d/three-graphics/dist/plugins/lights.js
"use strict";
/**
 * @author       Yannick Deubel (https://github.com/yandeu)
 * @copyright    Copyright (c) 2020 Yannick Deubel; Project Url: https://github.com/enable3d/enable3d
 * @license      {@link https://github.com/enable3d/enable3d/blob/master/LICENSE|GNU GPLv3}
 */
Object.defineProperty(exports, "__esModule", { value: true });
const index_1 = require("@enable3d/three-wrapper/dist/index");
const index_2 = require("@enable3d/three-wrapper/dist/index");
const index_3 = require("@enable3d/three-wrapper/dist/index");
class PointLightHelper extends index_1.Object3D {
    constructor(scene, light, size, color) {
        super();
        this.scene = scene;
        this.light = light;
        this.size = size;
        this.color = color;
        this.geo = new index_1.SphereBufferGeometry(size || 0.2, 16, 8);
        this.mat = new index_1.MeshBasicMaterial({ color: color || light.color });
        this.mesh = new index_1.Mesh(this.geo, this.mat);
        this.add(this.mesh);
        light.add(this);
    }
    dispose() {
        this.mesh.geometry.dispose();
        if (!Array.isArray(this.mesh.material))
            this.mesh.material.dispose();
        else
            this.mesh.material.forEach(m => m.dispose());
        this.remove(this.mesh);
    }
    update() { }
}
class Lights {
    constructor(scene) {
        this.scene = scene;
    }
    get helper() {
        return {
            directionalLightHelper: (light, size, color) => {
                const helper = new index_3.DirectionalLightHelper(light, size, color);
                this.scene.add(helper);
                return helper;
            },
            spotLightHelper: (light, color) => {
                const helper = new index_3.SpotLightHelper(light, color);
                this.scene.add(helper);
                return helper;
            },
            pointLightHelper: (light, size, color) => new PointLightHelper(this.scene, light, size, color)
        };
    }
    directionalLight(options = {}) {
        const { color = 0xffffff, intensity = 1 } = options;
        const light = new index_2.DirectionalLight(color, intensity);
        light.castShadow = true;
        this.scene.add(light);
        return light;
    }
    hemisphereLight(options = {}) {
        const { skyColor = 0xffffff, groundColor = 0xffffff, intensity = 1 } = options;
        const light = new index_2.HemisphereLight(skyColor, groundColor, intensity);
        this.scene.add(light);
        return light;
    }
    ambientLight(options = {}) {
        const { color = 0xffffff, intensity = 1 } = options;
        const light = new index_2.AmbientLight(color, intensity);
        this.scene.add(light);
        return light;
    }
    pointLight(options = {}) {
        const { color = 0xffffff, intensity = 1, distance = 0, decay = 1 } = options;
        const light = new index_2.PointLight(color, intensity, distance, decay);
        light.castShadow = true;
        this.scene.add(light);
        return light;
    }
    spotLight(options = {}) {
        const { color = 0xffffff, intensity = 1, distance = 0, angle = Math.PI / 8, penumbra = 0, decay = 1 } = options;
        const light = new index_2.SpotLight(color, intensity, distance, angle, penumbra, decay);
        light.castShadow = true;
        this.scene.add(light);
        return light;
    }
    rectAreaLight(options = {}) {
        const { color = 0xffffff, intensity = 1, width = 10, height = 10 } = options;
        const light = new index_2.RectAreaLight(color, intensity, width, height);
        this.scene.add(light);
        return light;
    }
}
exports.default = Lights;
