// @enable3d/three-graphics/dist/plugins/misc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TextureCube = void 0;
// misc
const water_1 = require("@enable3d/common/dist/misc/water");
const dist_1 = require("@enable3d/three-wrapper/dist");
class Misc {
    constructor(scene, renderer, factories) {
        this.scene = scene;
        this.renderer = renderer;
        this.factories = factories;
    }
    water(config = {}) {
        water_1.addWater(this.scene, this.renderer, config);
    }
    textureCube(textures) {
        if (textures.length !== 6) {
            console.log('[enable3d] You need to pass 6 urls to textureCube()');
        }
        const textureCube = new TextureCube();
        textures.forEach((texture, i) => {
            texture.wrapS = texture.wrapT = dist_1.RepeatWrapping;
            const material = this.factories.add.material({ phong: { map: texture } });
            textureCube.materials[i] = material;
        });
        return textureCube;
    }
}
exports.default = Misc;
class TextureCube {
    constructor() {
        this.materials = new Array(6);
    }
    get texture() {
        return {
            left: this.getTexture(0),
            right: this.getTexture(1),
            up: this.getTexture(2),
            down: this.getTexture(3),
            front: this.getTexture(4),
            back: this.getTexture(5)
        };
    }
    getTexture(i) {
        // @ts-ignore
        return this.materials[i].map;
    }
}
exports.TextureCube = TextureCube;
