// @enable3d/three-graphics/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MathUtils = exports.Vector3 = exports.Vector2 = exports.Object3D = exports.FBXLoader = exports.GLTFLoader = exports.RGBAFormat = exports.Texture = exports.WebGLRenderer = exports.Types = exports.THREE = exports.PhysicsLoader = exports.ThreeGraphics = void 0;
const core_1 = require("./core");
Object.defineProperty(exports, "ThreeGraphics", { enumerable: true, get: function () { return core_1.ThreeGraphics; } });
const physicsLoader_1 = require("@enable3d/common/dist/physicsLoader");
Object.defineProperty(exports, "PhysicsLoader", { enumerable: true, get: function () { return physicsLoader_1.PhysicsLoader; } });
const THREE = require("@enable3d/three-wrapper/dist/index");
exports.THREE = THREE;
const Types = require("@enable3d/common/dist/types");
exports.Types = Types;
const version_1 = require("./version");
var index_1 = require("@enable3d/three-wrapper/dist/index");
Object.defineProperty(exports, "WebGLRenderer", { enumerable: true, get: function () { return index_1.WebGLRenderer; } });
Object.defineProperty(exports, "Texture", { enumerable: true, get: function () { return index_1.Texture; } });
Object.defineProperty(exports, "RGBAFormat", { enumerable: true, get: function () { return index_1.RGBAFormat; } });
Object.defineProperty(exports, "GLTFLoader", { enumerable: true, get: function () { return index_1.GLTFLoader; } });
Object.defineProperty(exports, "FBXLoader", { enumerable: true, get: function () { return index_1.FBXLoader; } });
Object.defineProperty(exports, "Object3D", { enumerable: true, get: function () { return index_1.Object3D; } });
Object.defineProperty(exports, "Vector2", { enumerable: true, get: function () { return index_1.Vector2; } });
Object.defineProperty(exports, "Vector3", { enumerable: true, get: function () { return index_1.Vector3; } });
Object.defineProperty(exports, "MathUtils", { enumerable: true, get: function () { return index_1.MathUtils; } });
const info = `Powered by enable3d v${version_1.VERSION}`;
console.log(`%c %c %c %c %c ${info} %c https://enable3d.io/`, 'background: #ff0000', 'background: #ffff00', 'background: #00ff00', 'background: #00ffff', 'color: #fff; background: #000000;', 'background: none');
